### PR TITLE
Move everything over to typescript

### PR DESCRIPTION
**Pull request by brianwchou**
### Description
Migrating to typescript and adding types to files

a couple of things to note:
- react component files along with their tests should use the .tsx extension and all other regular files should use .ts. I believe this is for the typescript compiler to recognize jsx
- jest test coverage no longer works with the latest create-react-app scripts (i added a command to get coverage to work in a commit) as mentioned here https://github.com/facebook/create-react-app/issues/6888
- I've also done a forced audit on my machine and now npm shows 0 vulnerability warnings

- this is a good place to get started with placing types. Even when you don't know what the values are:
http://www.albertgao.xyz/2016/08/11/how-to-declare-a-function-type-variable-in-typescript/
- this confused me while researching this. If you want to use a javascript object as a hash table then its interface declaration should look something like this: 
`interface CardCount {
    [index: string]: number
}
`
You can find more about it here: https://www.typescriptlang.org/docs/handbook/interfaces.html
under indexable types

### CODE REVIEW

**Comment by brianwchou** [comment]
> so the type any is necessary in typescript?
any type is a default in typescript. Think about how we spoke about how all javascript types are implicitly objects, in typescript everything is by default any type.

> why are we migrating to typescript?
I wanted to see what all the hubbub was about. I spoke to a frontend engineer about the language and he highly recommended it for all the static typing protection. Also i looked into interfaces and i was checking out code. I don't remember what any of these objects in our code contain. I think having statements that explicitly say what objects are will help with development.

File: src/common/testData.ts [hunk 1]
```diff
@@ -20,4 +20,28 @@ export const cardInfo = {
     setName: undefined,
     toughness: "11",
     typeLine: "Artifact Creature — Golem",
+}
+
+interface CardInfo {
```
Comment (brianwchou) [inline src/common/testData.ts]: We probably don't need CardInfo inside this file. I think we can take this out

File: src/actions/SearchActions.ts [hunk 2]
```diff
@@ -9,7 +9,7 @@ export const SEARCH = {
 /*
     action creators
 */
-export const loadSearchCards = (cards) => {
+export const loadSearchCards = (cards: any) => {
```
Comment (brianwchou) [inline src/actions/SearchActions.ts]: Lets refrain from using any to the best of our understanding with typescript

Comment (zhc1289) [reply src/actions/SearchActions.ts]: Yes, we should scarcely use the type "any".

File: src/actions/CardActions.ts [hunk 3]
```diff
@@ -15,8 +16,8 @@ export const MAYBEBOARD = {
     REMOVE: "MAYBEBOARD_REMOVE",
 }
 
-export const addToDeckList = (cardInfo) => {
-    return (dispatch, getState) => {
+export const addToDeckList = (cardInfo: CardInfo) => {
```
Comment (brianwchou) [inline src/actions/CardActions.ts]: update CardInfo and avoid using any in this file

File: src/actions/SearchActions.ts (hunk 2 above)
Comment (brianwchou) [reply src/actions/SearchActions.ts]: no need to repeat what I've just said

File: src/components/Metrics/Graph.tsx [hunk 4]
```diff
@@ -18,7 +26,7 @@ export default function Graph({metrics}) {
 
     // sets up the vertical bars
     const renderBars = () => {
-        let sumOfAll = metrics.reduce((acc, cost) => {
+        let sumOfAll = metrics.reduce((acc: number, cost: any) => {
```
Comment (brianwchou) [inline src/components/Metrics/Graph.tsx]: again, refrain from any

File: src/containers/DeckList/DeckListEntry.tsx [hunk 5]
```diff
@@ -1,7 +1,15 @@
...
-    const handleOnClick = (e) => {
+interface DeckListEntryProps {
+    index: number,
+    card: CardInfo,
+    count: number,
+    getCardInfo(card: CardInfo, name: string): void   
+}
+
+export default function DeckListEntry ({index, card, count, getCardInfo}: DeckListEntryProps) {
+    const handleOnClick = (e: any) => {
```
Comment (brianwchou) [inline src/containers/DeckList/DeckListEntry.tsx]: I think any is okay here. don't really know what events are made up of, nor do i know how to handle them with typescript. Probably should look into this.

File: src/components/Card/Card.tsx [hunk 6]
```diff
@@ -1,15 +1,20 @@
 import React, { Component } from 'react';
 import './Card.css';
 
-
-export default class Card extends Component {
-    constructor(props) {
+export interface CardProps {
+    info: any,
+    buttonDisplay: any,
```
Comment (zhc1289) [inline src/components/Card/Card.tsx]: redefine type for buttonDisplay, which is a prop

File: src/containers/Search/Search.tsx [hunk 7]
```diff
@@ -1,6 +1,6 @@
 import React from 'react'
 import { connect } from 'react-redux';
-import { filteredSearchURL } from 'utility/URLs';
+import { filteredSearchURL } from 'common/URLs';
```
Comment (zhc1289) [inline src/containers/Search/Search.tsx]: renamed utility folder to common, fits the name better

File: tsconfig.json [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    "forceConsistentCasingInFileNames": true,
+    "module": "esnext",
+    "moduleResolution": "node",
+    "resolveJsonModule": true,
+    "isolatedModules": true,
+    "noEmit": true,
+    "jsx": "preserve"
+  },
+  "files": [
+      "/home/zhi/Desktop/MTG/deck-builder-web/src/components/Metrics/Graph.tsx",
```
Comment (brianwchou) [inline tsconfig.json]: this needs to be fixed

File: src/actions/CardActions.ts (hunk 3 above)
Comment (brianwchou) [reply src/actions/CardActions.ts]: I'm going to come back to this. Hard to find examples of how to handle dispatch with redux thunks

File: src/components/Card/Card.tsx (hunk 6 above)
Comment (brianwchou) [reply src/components/Card/Card.tsx]: ya

**Pull request merged**